### PR TITLE
exec_fd event source rework

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

File: src/core/service.c [hunk 1]
```diff
@@ -3391,6 +3391,11 @@ static void service_sigchld_event(Unit *u, pid_t pid, int code, int status) {
         else
                 clean_mode = EXIT_CLEAN_DAEMON;
 
+        if (s->main_pid == pid)
```
Comment (bluca) [inline src/core/service.c:3394]: If there are still processes active in the cgroup after main_pid has exited (`ExitType=cgroup`), would this cause issues? Or is it unrelated?

Comment (keszybz) [reply src/core/service.c:3394]: I think that's fine. By closing this even source, we are saying that we are not interested in the exec-fd event. But we know that this particular process has dies already, so we don't need the information from exec-fd anymore. Whether there are other process or not doesn't seem relevant.

**Review by bluca — APPROVED** [review]
Superficially LGTM, but not super familiar with this part.

**Comment by poettering** [comment]
We usually destory runtime objects that only make sense in some states from service_set_state(). Why wouldn't that work here too?

**Comment by keszybz** [comment]
> We usually destory runtime objects that only make sense in some states from service_set_state(). Why wouldn't that work here too?

Maybe? I like that we immediately destroy the resource when we learn that it's not valid anymore. I don't think doing it from `service_set_state()` gains us anything.

**Comment by keszybz** [comment]
I'll merge this, because we need *some* solution. There was a weak approval from @bluca… We know it at least fixes the issue and isn't obviously broken. We can always refactor it into something better later on.

**Comment by poettering** [comment]
> Maybe? I like that we immediately destroy the resource when we learn that it's not valid anymore. I don't think doing it from `service_set_state()` gains us anything.

well, from history we learned that binding destruction of objects to state instead of events is usually safer. i.e. if someone issues a daemon-reload while we are running, and changes the state Type=exec → Type=simple or so, then we should free the exec_fd too, but currently do not.

**Comment by poettering** [comment]
The thing is that service_set_state() already carries the destruction of the event source anyway. So why is the additional one still necessary that the last patch of this PR adds? I don't get it?

**Comment by keszybz** [comment]
Dunno, this code is so freakishly complicated that it's hard to say for certain. The issue was still reproducible, so it was certainly broken with before the patch.

**Comment by keszybz** [comment]
From the commit message:

    service_enter_stop_post() calls service_set_state(), which will also destroy
    the source. But that happens too late, because from service_enter_stop_post()
    we call service_spawn() first, and then service_set_state() second.

Is this answer enough?

I'll file a follow-up patch to expand the comment and move the unref a few lines down into the other if.

**Comment by poettering** [comment]
Ah, excellent! thanks.

Yeah, adding this as comment to the sources would be perfect indeed! thanks for looking into this.

**Pull request merged**